### PR TITLE
Document uninstallation, backup and restore

### PR DESCRIPTION
**Pull request by adrianludwin**
Also update some unclear Gitops directions.

### CODE REVIEW

**Comment by adrianludwin** [comment]
/assign @yiqigao217 
/assign @rjbez17 

This is a bit subtle; does this look good enough for now?

cc @sophieliu15 , @adrianchung

File: incubator/hnc/docs/user-guide/how-to.md [hunk 1]
```diff
@@ -388,19 +389,70 @@ make deploy
...
-kubectl delete -f https://github.com/kubernetes-sigs/multi-tenancy/releases/download/hnc-${HNC_VERSION}/hnc-manager.yaml
+kubectl -n hnc-system delete deployment hnc-controller-manager
+kubectl delete validatingwebhookconfiguration.admissionregistration.k8s.io hnc-validating-webhook-configuration
+```
+
+You may also completely delete HNC, including its CRDs and namespaces. However,
+**this is a destructive process that results in data loss.** In particular, you
+will lose any cluster-wide configuration in your `HNCConfig` object, as well as
+any hierarchical relationships between different namespaces, _excluding_
+subnamespaces (subnamespace relationships are saved as annotations on the
```
Comment (adrianludwin) [inline incubator/hnc/docs/user-guide/how-to.md:403]: Note that this will only be true after #762 is merged.

File: incubator/hnc/docs/user-guide/how-to.md [hunk 2]
```diff
@@ -388,19 +389,70 @@ make deploy
...
+to delete the `.metadata.uid` and `metadata.resourceVersion` fields from each
+object **prior** to applying the file.  It may also be convenient to delete the
+`.metadata.selfLink` field, which is alphabetically between the other two
+fields; this is safe.
+
+Once you are ready, first reapply the cluster-wide configuration via `kubectl
+apply -f hncconfig.yaml`, and then the structural relationships via `kubectl
+apply -f structure.yaml`.
+
+Finally, resolve any `SubnamespaceAnchorMissing` conditions. Type `kubectl hns
```
Comment (adrianludwin) [inline incubator/hnc/docs/user-guide/how-to.md:451]: This requires #766 to be merged.

File: incubator/hnc/docs/user-guide/best-practices.md [hunk 3]
```diff
@@ -75,20 +75,30 @@ kubectl hns describe foo
...
+`HierarchyConfiguration` objects in each namespace), and not your subnamespaces
+(the `SubnamespaceAnchor` objects) since subnamespaces are mainly for
+unprivileged users, not Gitops flows.
+
+When applying your `HierarchyConfiguration` objects, HNC imposes two
+restrictions on the order in which changes can be applied:
 
 1. A namespace must exist before it can be referenced as the parent of another
-   namespace. However, the would-be parent’s `HierarchicalConfiguration` does
-   _not_ need to exist.
```
Comment (yiqigao217) [inline incubator/hnc/docs/user-guide/best-practices.md:83]: Why is this removed?

**Comment by adrianludwin** [comment]
> Also do we need to mention that not only the hierarchy will be gone but also the propagated objects?

That's a good point. I'm wondering if we should prevent HNC from removing the propagated objects if the CRDs get deleted. After all, if the pod dies first, everything will stay in place; if the CRDs _start_ getting deleted, _then_ the pod dies, then an unknown number of objects will be left around. Do you agree?

File: incubator/hnc/docs/user-guide/best-practices.md (hunk 3 above)
Comment (adrianludwin) [reply incubator/hnc/docs/user-guide/best-practices.md:83]: It just seemed redundant to me - I didn't think it would be a question that people would ask.

Comment (adrianludwin) [reply incubator/hnc/docs/user-guide/best-practices.md:83]: wdyt?

Comment (yiqigao217) [reply incubator/hnc/docs/user-guide/best-practices.md:83]: sg

**Comment by yiqigao217** [comment]
> I'm wondering if we should prevent HNC from removing the propagated objects if the CRDs get deleted. After all, if the pod dies first, everything will stay in place; if the CRDs _start_ getting deleted, _then_ the pod dies, then an unknown number of objects will be left around. Do you agree?

Agreed. Just confirming we don't have this prevention right now, right?


> @yiqigao217 We don't put finalizers on anything other than `SubnamespaceAnchors` for cascading deletion, right? So if all anchors are deleted (because the CRD is removed), there should be nothing stopping users from deleting their namespaces?

We also put finalizers on `HierarchyConfiguration` objects if the namespace has any subnamespace. Therefore, subnamespace can be deleted by users, but any namespace (including subnamespaces) with subnamespaces cannot be deleted.

**Comment by adrianludwin** [comment]
> > I'm wondering if we should prevent HNC from removing the propagated objects if the CRDs get deleted. After all, if the pod dies first, everything will stay in place; if the CRDs _start_ getting deleted, _then_ the pod dies, then an unknown number of objects will be left around. Do you agree?
> 
> Agreed. Just confirming we don't have this prevention right now, right?

That's right - we don't.

> We also put finalizers on `HierarchyConfiguration` objects if the namespace has any subnamespace. Therefore, subnamespace can be deleted by users, but any namespace (including subnamespaces) with subnamespaces cannot be deleted. But if the `HierarchyConfiguration` CRD is also removed, there should be nothing stopping users from deleting any namespaces.

Ok, I just tried this and it works. Thanks! Fortunately, it looks like deleting a CRD isn't blocked by finalizers.

**Comment by yiqigao217** [comment]
> > We also put finalizers on `HierarchyConfiguration` objects if the namespace has any subnamespace. Therefore, subnamespace can be deleted by users, but any namespace (including subnamespaces) with subnamespaces cannot be deleted. But if the `HierarchyConfiguration` CRD is also removed, there should be nothing stopping users from deleting any namespaces.
> 
> Ok, I just tried this and it works. Thanks! Fortunately, it looks like deleting a CRD isn't blocked by finalizers.

Nice!

/lgtm
/approve
/hold
/assign @rjbez17

**Comment by adrianludwin** [comment]
@rjbez17 lmk if you have any concerns and I'll address them in a followup PR.
